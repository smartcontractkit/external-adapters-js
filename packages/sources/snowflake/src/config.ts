import { Requester, util } from '@chainlink/ea-bootstrap'
import { Config } from '@chainlink/types'

export const NAME = 'SNOWFLAKE'

export const DEFAULT_ENDPOINT = 'covid-cases'
export const DEFAULT_DATABASE = 'COVID19_BY_STARSCHEMA_DM'
export const DEFAULT_SCHEMA = 'PUBLIC'

export interface SnowflakeConfig extends Config {
  account: string
  baseURL: string
  database: string
  privateKey: string
  qualifiedUsername: string
  schema: string
  username: string
}

export const makeConfig = (prefix?: string): SnowflakeConfig => {
  const account: string = util.getRequiredEnv('ACCOUNT')
  const privateKey: string = util.getRequiredEnv('PRIVATE_KEY')
  const username: string = util.getRequiredEnv('DB_USERNAME')

  const database: string = util.getEnv('DATABASE') || DEFAULT_DATABASE
  const schema: string = util.getEnv('SCHEMA') || DEFAULT_SCHEMA

  const provider: string | undefined = util.getEnv('CLOUD_PROVIDER')
  const region: string | undefined = util.getEnv('CLOUD_REGION')

  const fullAccount = [account, region, provider].filter((i) => !!i).join('.')
  const qualifiedUsername = `${account.toUpperCase()}.${username.toUpperCase()}`
  const baseURL = `https://${fullAccount}.snowflakecomputing.com/api`

  const config = {
    ...Requester.getDefaultConfig(prefix),
    account,
    baseURL,
    database,
    privateKey,
    qualifiedUsername,
    schema,
    username,
  }

  config.defaultEndpoint = DEFAULT_ENDPOINT

  return config
}
