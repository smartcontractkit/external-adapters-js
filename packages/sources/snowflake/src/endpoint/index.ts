import type { TInputParameters as CovidCasesInputParameters } from './covid-cases'

export type TInputParameters = CovidCasesInputParameters

export * as covidCases from './covid-cases'
