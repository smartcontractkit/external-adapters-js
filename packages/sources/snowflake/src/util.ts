import crypto from 'crypto'
import jwt from 'jsonwebtoken'

const LIFETIME = 120 * 1000 // milliseconds
const ALGORITHM = 'RS256'
const ISSUER = 'iss'
const SUBJECT = 'sub'
const EXPIRE_TIME = 'exp'
const ISSUE_TIME = 'iat'

export function buildSnowflakeJWT(config: {
  privateKey: string
  qualifiedUsername: string
}): string {
  const publicKey = crypto
    .createPublicKey({
      key: config.privateKey,
      format: 'pem',
    })
    .export({
      format: 'der',
      type: 'spki',
    })

  const publicKeyFingerprint =
    'SHA256:' + crypto.createHash('sha256').update(publicKey).digest('base64')

  const currentTime = Date.now()
  const jwtTokenExp = currentTime + LIFETIME

  const jwtPayload = {
    [ISSUER]: `${config.qualifiedUsername}.${publicKeyFingerprint}`,
    [SUBJECT]: `${config.qualifiedUsername}`,
    [ISSUE_TIME]: currentTime,
    [EXPIRE_TIME]: jwtTokenExp,
  }

  return jwt.sign(jwtPayload, config.privateKey, { algorithm: ALGORITHM })
}
