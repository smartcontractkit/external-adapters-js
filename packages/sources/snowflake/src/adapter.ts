import { Builder } from '@chainlink/ea-bootstrap'
import {
  ExecuteWithConfig,
  ExecuteFactory,
  AdapterRequest,
  APIEndpoint,
} from '@chainlink/ea-bootstrap'
import { makeConfig, SnowflakeConfig } from './config'
import * as endpoints from './endpoint'

export const execute: ExecuteWithConfig<SnowflakeConfig, endpoints.TInputParameters> = async (
  request,
  context,
  config,
) => {
  return Builder.buildSelector<SnowflakeConfig, endpoints.TInputParameters>(
    request,
    context,
    config,
    endpoints,
  )
}

export const endpointSelector = (
  request: AdapterRequest,
): APIEndpoint<SnowflakeConfig, endpoints.TInputParameters> =>
  Builder.selectEndpoint<SnowflakeConfig, endpoints.TInputParameters>(
    request,
    makeConfig(),
    endpoints,
  )

export const makeExecute: ExecuteFactory<SnowflakeConfig, endpoints.TInputParameters> = (
  config,
) => {
  return async (request, context) => execute(request, context, config || makeConfig())
}
