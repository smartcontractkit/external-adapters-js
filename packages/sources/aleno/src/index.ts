import { expose, ServerInstance } from '@chainlink/external-adapter-framework'
import { Adapter } from '@chainlink/external-adapter-framework/adapter'
import { config } from './config'
import { price } from './endpoint'

export const adapter = new Adapter({
  defaultEndpoint: price.name,
  name: 'ALENO',
  config,
  endpoints: [price],
  rateLimiting: {
    tiers: {
      default: {
        rateLimit1m: 30,
      },
    },
  },
})

export const server = (): Promise<ServerInstance | undefined> => expose(adapter)
