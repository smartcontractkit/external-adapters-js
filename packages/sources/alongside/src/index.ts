import { expose, ServerInstance } from '@chainlink/external-adapter-framework'
import { Adapter } from '@chainlink/external-adapter-framework/adapter'
import { collateral } from './endpoint'
import { config } from './config'

export const adapter = new Adapter({
  defaultEndpoint: collateral.name,
  name: 'ALONGSIDE',
  config,
  endpoints: [collateral],
})

export const server = (): Promise<ServerInstance | undefined> => expose(adapter)
