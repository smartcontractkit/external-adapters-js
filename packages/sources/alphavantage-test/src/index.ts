import { expose, ServerInstance } from '@chainlink/external-adapter-framework'
import { PriceAdapter } from '@chainlink/external-adapter-framework/adapter'
import { config } from './config'
import { forex } from './endpoint'

export const adapter = new PriceAdapter({
  defaultEndpoint: forex.name,
  name: 'ALPHAVANTAGE',
  config,
  endpoints: [forex],
  rateLimiting: {
    tiers: {
      free: {
        rateLimit1m: 5,
        rateLimit1h: 20.83,
      },
      '49.99': {
        rateLimit1m: 75,
      },
      '99.99': {
        rateLimit1m: 150,
      },
      '149.99': {
        rateLimit1m: 300,
      },
      '199.99': {
        rateLimit1m: 600,
      },
      '249.99': {
        rateLimit1m: 1200,
      },
    },
  },
})

export const server = (): Promise<ServerInstance | undefined> => expose(adapter)
