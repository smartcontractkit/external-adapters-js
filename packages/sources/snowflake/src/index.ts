import { expose } from '@chainlink/ea-bootstrap'
import { makeExecute, endpointSelector } from './adapter'
import { makeConfig, NAME } from './config'

const { server } = expose(NAME, makeExecute(), undefined, endpointSelector)
export { NAME, makeExecute, makeConfig, server }
