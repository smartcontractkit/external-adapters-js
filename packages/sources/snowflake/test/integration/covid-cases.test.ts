import { AdapterRequest } from '@chainlink/ea-bootstrap'
import process from 'process'
import { generateKeyPair } from 'crypto'
import { server as startServer } from '../../src'
import { mockSnowflakeResponse } from './fixtures'
import { setupExternalAdapterTest } from '@chainlink/ea-test-helpers'
import type { SuiteContext } from '@chainlink/ea-test-helpers'
import { SuperTest, Test } from 'supertest'

async function generatePrivateKey(): Promise<string> {
  return new Promise((resolve) => {
    generateKeyPair(
      'rsa',
      {
        modulusLength: 4096,
        publicKeyEncoding: {
          type: 'spki',
          format: 'pem',
        },
        privateKeyEncoding: {
          type: 'pkcs8',
          format: 'pem',
        },
      },
      (_, __, privateKey) => {
        resolve(privateKey)
      },
    )
  })
}

beforeAll(async () => {
  process.env.PRIVATE_KEY = process.env.PRIVATE_KEY ?? (await generatePrivateKey())
})

describe('execute', () => {
  const id = '1'
  const context: SuiteContext = {
    req: null,
    server: startServer,
  }

  const envVariables = {
    CACHE_ENABLED: 'false',
    ACCOUNT: process.env.ACCOUNT ?? 'test_account',
    DB_USERNAME: process.env.DB_USERNAME ?? 'test_db_username',
    DATABASE: process.env.DATABASE ?? 'test_database',
    SCHEMA: process.env.SCHEMA ?? 'test_schema',
    CLOUD_REGION: process.env.CLOUD_REGION ?? 'test_region',
    CLOUD_PROVIDER: process.env.CLOUD_PROVIDER ?? 'test_provider',
    API_VERBOSE: true as unknown as string,
  }

  setupExternalAdapterTest(envVariables, context)

  describe('basic', () => {
    const data: AdapterRequest = {
      id,
      data: {
        state: 'Alabama',
        county: 'Autauga',
      },
    }

    it('should return success', async () => {
      mockSnowflakeResponse()

      const response = await (context.req as SuperTest<Test>)
        .post('/')
        .send(data)
        .set('Accept', '*/*')
        .set('Content-Type', 'application/json')
        .expect('Content-Type', /json/)
        .expect(200)
      expect(response.body).toMatchSnapshot()
    })
  })
})
