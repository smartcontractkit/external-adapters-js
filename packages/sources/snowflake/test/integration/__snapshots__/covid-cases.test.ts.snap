// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`execute basic should return success 1`] = `
{
  "data": {
    "code": "090001",
    "createdOn": 1638467668319,
    "data": [
      [
        "10531",
      ],
    ],
    "message": "Statement executed successfully.",
    "requestId": "e68f6b77-72bd-4b7f-8719-247ac70f6f82",
    "result": 10531,
    "resultSetMetaData": {
      "format": "jsonv2",
      "numRows": 1,
      "partitionInfo": [
        {
          "rowCount": 1,
          "uncompressedSize": 9,
        },
      ],
      "rowType": [
        {
          "byteLength": null,
          "collation": null,
          "database": "test_database",
          "length": null,
          "name": "CONFIRMED",
          "nullable": true,
          "precision": 38,
          "scale": 0,
          "schema": "test_schema",
          "table": "JHU_DASHBOARD_COVID_19_GLOBAL",
          "type": "fixed",
        },
      ],
    },
    "sqlState": "00000",
    "statementHandle": "01a0af12-0000-5d4d-0000-000131664085",
    "statementStatusUrl": "/api/statements/01a0af11-0000-5d4e-0000-0001316650a5?requestId=e68f6b77-72bd-4b7f-8719-247ac70f6f82",
  },
  "jobRunID": "1",
  "providerStatusCode": 200,
  "result": 10531,
  "statusCode": 200,
}
`;
